README.md
# EzerCloud

Welcome to EzerCloud!
This server is based on ruby-on-rails (RoR) web framework and serves data collection/surveillance/analytics for health data transferred from various health device like FREND, BUDDI, and EzerReader and for surveys data from online survey system like Commcare HQ.

## Requirements

1. PostgresSQL DBMS (>= 9.5)
We are developing on PostgresSQL DBMS. Although other DBMS could be applicable, we do not recommended.

2. Elasticsearch (>= 5.0.1)

3. Redis (>= 3.2.5)

4. Ruby (>= 2.3.1) and RoR 5.0.0.1

#### Optional requirements

1. R (>=3.3.2)

2. TeX Live (>= 2016)

## Contributing
1. Fork it (http://bitbucket/skonmeme/ezercloud/fork)

2. Create your feature branch (git checkout -b my-new-feature)

3. Commit your changes (git commit -am 'Add some feature')

4. Push to the branch (git push origin my-new-feature)

5. Create new Pull Request

CHANGELOG.md
  Note: This is in reverse chronological order, so newer entries are added to the top.

## Skynet 0.3

######2016. 12. 9
Diagnosis
* a bug that the transaction to save a diagnosis and a measurement had not worked correctly was corrected. In transaction routine, 'raise ActiveRecord::Rollback' should be called to roll back.
* routines to set additional variables were moved to Diagnosis model.
* a device_license registration should be skipped when user_id is not related with account.

######2016. 12. 1
System
* QR code & Barcode generator gems were added.
* controller routes would be limited only to permitted pages gradually.

Commcare
* Commcare model was added to store survey data from CommcareHQ.
* lib/taks/commcare.rake will perform a task to contact and take new surveys periodically. Currenty 'Student Survey' forms would be taken. 

CommcareImage
* images will be stored from CommcareHQ surveys.

Diagnosis
* subject_id was added to make a link to a subject (Person model) to be taken an examine. If there is not subject information yet, new subject will be made automatically.
* when user_id exists and links to an account, the account will be checked whether owned device_licent is issued or not about this device. new device_license will be issued automatically if device_license is not issued.

Person
* Person model is set up in detail.
* the title of view pages were located in 'Health Records'.

Buddi/EzerCloud/Frend
* Reindexing device information will be performed after new device is registered, which is utilized to search serial numbers of devices quickly.

Account/Management, Account/Role
* some view pages were renewed to show preciser information.

######2016. 11. 21
DiagnosisImage
* wrong validation step was removed.

######2016. 11. 18
System
* bootstrap-datetimepicker gem replaced bootstrap-datepicker3 gem.

Device/Equipment/ErrorCode/Laboratory
* accounts with :data_manager or :device_manager are only accesible.

Diagnosis
* techinican instance variable is also securely stored.
* equipment instance variable should not be blank and creation request for invalid equipment returns fail.
* tranaction acts on a single database connection. For example, there is a redirection, transaction does not work correctly. In this situation, ActiveRecord::Rollback should be called manually. As a solution, we divided DB creation section and redirection section and applied transaction to DB creation section.

Notification
* NotificationError initialization bugs were fixed.
* a bug to generate exception was removed when there is no information, which should be ignored without any exception.

Person
* authorization should be required like Diagnosis (need validation).

Buddi/EzerReader/Frend
* devices should be accesible by diagnoses_controller only.

######2016. 11. 17
Diagnosis
* diagnosis validation requires remote_ip, which checks connected client IP. This validation is always failed because remote_ip is nil in migration.

Pundit
* EzerReaderPolicy and BuddiPolicy should be inherited from MachinePolicy.

######2016. 11. 16
System
* bootstrap-datepicker-rails gem was replaced by boostrap-datetimepicker-rails gem.
* jquery javascript was replaced by jquery3.
* all (almost) 'and' and 'or' were replaced by '&&' and '||' operators, respectively. 'and' and 'or' operators were strictly recommanded to be banned to clearify ruby coding style.
* addition blocks, :before_header and :after_footer were added in application layout.

Account/Device
* the bug that process performed with null device_license instance was fixed.
* test nil for null tag on @device_license (not nil but '') was fixed when an administrator grants directly.
* decline action name was changed to terminate (deactivate).
* redirection page was changed to modified account (for administrator) after termination.

Account/Management
* bugs on routes to remove roles or to terminate devices were fixed.

Account/Role
* bootstrap-datepicker was replaced by bootstrap-datetimepicker.
* the bug that process performed with null role instance was fixed.
* test nil for null tag on @@role (not nil but '') was fixed when an administrator grants directly.
* redirection page was changed to modified account (for administrator) after destroy.
* Google Place should be loaded syncronously to work with other google place based javascripts correctly and reloaded every time without caching. To archive, asyncronous (with deferred) javascript loading was removed and turbolinks were disabled (link with data: { turbolinks: false } parameter).

Google Map
* Google Map and Google Place should be loaded syncronously to work with other google map/place based javascripts correctly and reloaded every time without caching. To archive, asyncronous (with deferred) javascript loading was removed and turbolinks were disabled (link with data: { turbolinks: false } parameter).

######2016. 11. 14
System
* All types migration errors were removed on clean (empty) DB (should be careful for DB migration fail), which does not guarantee DB migration on used DB (durty DB?).
* Welcome page is prepared for front page (before and after login).

Pundit
* Authorization 
* Roll settings are connected with rollify gem.

Diagnosis
* Pubdit policy is applied with :data_manager
* device owner could access own measurements.

Google Map/AssayKit/QualityControlMaterial/Report
* Pubdit policy is applied with :data_manager.
* device owner could access surviallence map only for own measurements (not yet).

Role
* Rollify gem has been applied, which includes creating, requesting, applying, and removing roles.
* accounts/management_controller shows aquired roles. 

Device & DeviceLicense
* Device class manages collected devices information, and Accounts with related DeviceLicense can only access measurement data.
* accounts/management_controller shows aquired device_licenses.
* Device could be registered automatically after related measurement data has been registered.

db/seeds.rb
# This file should contain all the record creation needed to seed the database with its default values.
# The data can then be loaded with the rake db:seed (or created alongside the db with db:setup).
#
# Examples:
#
#   cities = City.create([{ name: 'Chicago' }, { name: 'Copenhagen' }])
#   Mayor.create(name: 'Emanuel', city: cities.first)

# Desirable Biological Variation Database specifications: https://www.westgard.com/biodatabase1.htm

#u = User.create(name: "admin", password: "11112222", full_name: "Administrator", privilege_super: true, email: "skonmeme@gmail.com", authorized: true)
#u = User.create(name: "reagent", password: "22221111", full_name: "Reagent Manager for FREND", email: "skonmemekr@gmail.com", privilege_reagent: true, privilege_notification: true, equipment_frends: true, authorized: true)

# Equipment

e_frend = Equipment.create(equipment: 'FREND', manufacturer: 'NanoEnTek', klass: 'Frend', db: 'frends', tests: 3, prefix: 'test_')
e_buddi = Equipment.create(equipment: 'BUDDI', manufacturer: 'NanoEnTek', klass: 'Buddi', db: 'buddis', tests: 6, prefix: 'test_')
e_ezer_reader = Equipment.create(equipment: 'EzerReader', manufacturer: 'Unknown', klass: 'EzerReader', db: 'ezer_readers', tests: 6, prefix: 'test_')

# Assay kits & QC materials

r11 = Reagent.create(equipment: 'FREND', name: 'QC', number: 11, unit: 'NA')
r12 = Reagent.create(equipment: 'FREND', name: 'Troponin I', number: 12, unit: 'ng/mL', lod: 0.05, uod: 20, threshold: 0.4)
r13 = Reagent.create(equipment: 'FREND', name: 'CK-MB', number: 13, unit: 'ng/mL', lod: 1, uod: 80, threshold: 5)
r14 = Reagent.create(equipment: 'FREND', name: 'Myoglobin', number: 14, unit: 'ng/mL', lod: 5, uod: 500, threshold: 100)
r16 = Reagent.create(equipment: 'FREND', name: 'hsCRP', number: 16, unit: 'ng/mL', lod: 0.3, uod: 10, threshold: 1)
r17 = Reagent.create(equipment: 'FREND', name: 'D-Dimer', number: 17, unit: 'ng/mL', lod: 0, uod: 3000, threshold: 300)
r18 = Reagent.create(equipment: 'FREND', name: 'NT-proBNP', number: 18, unit: 'pg/mL', lod: 5, uod: 35000, threshold: 125)
r19 = Reagent.create(equipment: 'FREND', name: 'BNP', number: 19, unit: 'pg/mL', lod: 30, uod: 2500, threshold: 125)
r20 = Reagent.create(equipment: 'FREND', name: 'FABP', number: 20, unit: 'ng/mL', lod: 2, uod: 3000, threshold: 6)
r30 = Reagent.create(equipment: 'FREND', name: 'PSA', number: 30, unit: 'ng/mL', lod: 0.1, uod: 25, threshold: 4)
r32 = Reagent.create(equipment: 'FREND', name: 'AFP', number: 32, unit: 'ng/mL', lod: 1, uod: 400, threshold: 20)
r34 = Reagent.create(equipment: 'FREND', name: 'CEA', number: 34, unit: 'ng/mL', lod: 1, uod: 400, threshold: 10)
r35 = Reagent.create(equipment: 'FREND', name: 'Testosterone', number: 35, unit: 'ng/dL', lod: 20, uod: 1500, threshold: 10)
r40 = Reagent.create(equipment: 'FREND', name: 'TSH', number: 40, unit: 'mIU/L', lod: 0.06, uod: 25, threshold: 3.33)
r41 = Reagent.create(equipment: 'FREND', name: 'Free T4', number: 41, unit: 'ng/dL', lod: 0.4, uod: 6, threshold: 1)
r42 = Reagent.create(equipment: 'FREND', name: 'Total T3', number: 42, unit: 'ng/mL', lod: 0.25, uod: 8, threshold: 1)
r43 = Reagent.create(equipment: 'FREND', name: 'Free T3', number: 43, unit: 'pg/mL', lod: 1, uod: 30, threshold: 1)
r44 = Reagent.create(equipment: 'FREND', name: 'Total T4', number: 44, unit: 'nmol/L', lod: 12.87, uod: 308.88, threshold: 1)
r50 = Reagent.create(equipment: 'FREND', name: 'Vitamin D', number: 50, unit: 'ng/mL', lod: 10, uod: 110, threshold: 20)
r60 = Reagent.create(equipment: 'FREND', name: 'PCT', number: 60, unit: 'ng/mL', lod: 0.07, uod: 32, threshold: 1)
r70 = Reagent.create(equipment: 'FREND', name: 'SAA', number: 70, unit: 'mg/L', lod: 5, uod: 150, threshold: 10)
r75 = Reagent.create(equipment: 'FREND', name: 'Hp', number: 75, unit: 'mg/dL', lod: 30, uod: 400, threshold: 50)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'QC', kit: 11)
p = r11.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'Cardiac Triple', kit: 15)
p = r12.plates.create(assay_kit: a)
p = r13.plates.create(assay_kit: a)
p = r14.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'hsCRP', kit: 16)
p = r16.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'D-Dimer', kit: 17)
p = r17.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'NT-proBNP', kit: 18)
p = r18.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'BNP', kit: 19)
p = r19.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'FABP', kit: 20)
p = r20.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'Cardiac Duo', kit: 21)
p = r12.plates.create(assay_kit: a)
p = r19.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'PSA plus', kit: 30)
p = r30.plates.create(assay_kit: a)
p.quality_control_materials.create(service: 'BIO-RAD', lot: '40881', expire: Date.new(2017, 2, 28), equipment: 'FREND', manufacturer: 'NanoEnTek', mean: 0.270, sd: (0.270 - 0.190) / 3)
p.quality_control_materials.create(service: 'BIO-RAD', lot: '40882', expire: Date.new(2017, 2, 28), equipment: 'FREND', manufacturer: 'NanoEnTek', mean: 3.95, sd: (3.95 - 3.37) / 3)
p.quality_control_materials.create(service: 'BIO-RAD', lot: '40883', expire: Date.new(2017, 2, 28), equipment: 'FREND', manufacturer: 'NanoEnTek', mean: 25, sd: 0)

#a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'PAC', kit: 31)
#p = r30.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'AFP', kit: 32)
p = r32.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'CEA', kit: 34)
p = r34.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'Testosterone', kit: 35)
p = r35.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'TSH', kit: 40)
p = r40.plates.create(assay_kit: a)
p.quality_control_materials.create(service: 'BIO-RAD', lot: '40881', expire: Date.new(2017, 2, 28), equipment: 'FREND', manufacturer: 'NanoEnTek', mean: 0.630, sd: (0.630 - 0.440) / 3)
p.quality_control_materials.create(service: 'BIO-RAD', lot: '40882', expire: Date.new(2017, 2, 28), equipment: 'FREND', manufacturer: 'NanoEnTek', mean: 5.99, sd: (5.99 - 5.08) / 3)
p.quality_control_materials.create(service: 'BIO-RAD', lot: '40883', expire: Date.new(2017, 2, 28), equipment: 'FREND', manufacturer: 'NanoEnTek', mean: 25, sd: 0)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'Free T4', kit: 41)
p = r41.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'Total T3', kit: 42)
p = r42.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'Free T3', kit: 43)
p = r43.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'Total T4', kit: 44)
p = r44.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'Thyroid Duo', kit: 45)
p = r41.plates.create(assay_kit: a)
p = r40.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'Vitamin D', kit: 50)
p = r50.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'PCT', kit: 60)
p = r60.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'SAA', kit: 70)
p = r70.plates.create(assay_kit: a)

a = AssayKit.create(equipment: 'FREND', manufacturer: 'NanoEnTek', device: 'Hp', kit: 75)
p = r75.plates.create(assay_kit: a)

# Labaoratories
l = Laboratory.create(ip_address: '211.210.64.61', equipment: 'FREND', kit: 30)
l = Laboratory.create(ip_address: '211.210.64.61', equipment: 'FREND', kit: 40)
l = Laboratory.create(ip_address: '58.226.94.194', equipment: 'FREND', kit: 30)
l = Laboratory.create(ip_address: '58.226.94.194', equipment: 'FREND', kit: 40)

# Error codes for FREND

c = e_frend.error_codes.create(error_code: 'EMCI-01', description: 'The cartridge is not inserted.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMCI-02', description: 'The cartridge is not inserted completely.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMCI-03', description: 'The cartridge is inserted in the oppsite direction.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMCI-04', description: 'The cartridge is inserted upside-down.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMBR-01', description: 'The barcode printed on the cartridge may have been damaged so the reader is unable to interpret the code correctly.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMBR-02', description: 'The barcode information related to the cartridge lot number printed on the cartridge is not readable by the system reader.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMBR-03', description: 'The lot number entered for the cartridge does not match the lot number on the code chip.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMDB-01', description: 'The USB drive version is not correct, or the USB drive is not proerly connected to the FREND(tm) System. In addition, the USB drive or USB H port has somehow been damaged.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMDB-02', description: 'A problem has been detected when attempting to transfer data from the FREND(tm) System to the USB drive.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMDS-01', description: 'The storage capacity of the FREND(tm) System to store the data has been exceed.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMCC-01', description: 'The code chip is not inserted properly, the code chip has somehow been damaged or the FREND(tm) System does not recognize the code chip.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMCC-02', description: 'The FREND(tm) System does not recognize the code chip.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMTF-01', description: 'The lot number of the inserted cartridge is different from the lot number of the installed code chip.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMTF-02', description: 'Unable to output a test result the reference signal is weaker than the baseline.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMTF-03', description: 'Unable to output a test result because the reference signal is stronger than the baseline.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMTF-04', description: 'The system cannot detect any signal.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMTF-05', description: 'The system detects more than four signals because the cartridge is damaged.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMTF-06', description: 'The lateral flow of the sample within the cartridge is incomplete.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMTF-07', description: 'Time-out error', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMTF-08', description: 'The cartridge has passed the expiration date', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMUF-01', description: 'There is no executable file (FREND.exe) on the USB drive.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMUF-02', description: 'There is no executable file (FRENDUPDATE.exe) on the USB drive.', level: 'General')
c = e_frend.error_codes.create(error_code: 'EMER-01', description: 'The entry requirements are not entered.', level: 'General')
